### PR TITLE
Change `HTTP::Request#remote_address` type to `Socket::Address?`

### PR DESCRIPTION
**Pull request by waj**
This is a breaking change, but probably with a very low impact.

Currently the `remote_address` is a `String` with the format "ip:port", so if just the IP is needed extra parsing must be done. Also, this is consistent with other occurrences of `local_address` and `remote_address` in the std.

### CODE REVIEW

**Comment by asterite** [comment]
I implemented it as a string because in Go it's a string. I'm sure there must be a good reason why the chose String. Why do you need the address?

**Comment by waj** [comment]
In this case for logging just the IP. The port changes on every request and is not useful at all for 99.99% of the cases.

Other useful cases are auditing, filtering, geolocation of clients, stats. All these cases need just the IP address without the port.

**Comment by asterite** [comment]
Can you split the string by `:` and just keep the left-hand side?

I think this PR might be fine, but I'd like to know why Go chose to use a String.

**Comment by straight-shoota** [comment]
I believe a string is just more flexible.

> The port changes on every request and is not useful at all for 99.99% of the cases.

In light of that we might perhaps consider only storing the IP address in `remote_address`. That would need to be a string because we have no data type for only the IP without port.

**Comment by waj** [comment]
Go is not the only language out there 🙂 
  * Ruby: Rack has a `request.ip`.
  * Dart: https://api.dart.dev/stable/2.8.1/dart-io/HttpConnectionInfo-class.html
  * .NET: [RemoteIpAddress](https://docs.microsoft.com/en-us/dotnet/api/microsoft.net.http.server.request.remoteipaddress?view=aspnetcore-1.1#Microsoft_Net_Http_Server_Request_RemoteIpAddress)

**Comment by waj** [comment]
> I believe a string is just more flexible

Of course. What's more flexible than a string? But we don't put strings everywhere for that reason.  And the address can be easily converted to a string.

**Comment by waj** [comment]
> Can you split the string by `:` and just keep the left-hand side?

Of course I can parse the value, but first it has to be matched with a regex in case is not an `IPAddress`. But why add that overhead when the original `IPAddress` was already known?

In the future, if we find another possibility for `Address` that doesn't fit within the concept of a `Socket` we could move that class to a different, more general module and add another subclass.

**Comment by waj** [comment]
FYI, I just tested the performance of `samples/http_server.cr` in my machine with `wrk` and I obtain 130k req/s without the change and 139k req/s with it. Not a big difference specially thinking that this "hello world" app is not representative but to see that there is an impact creating extra strings.

**Comment by asterite** [comment]
Cool! If other languages provide a non-string value then this change is perfect 💯

File: src/http/server/handlers/log_handler.cr [hunk 1]
```diff
@@ -24,7 +24,18 @@ class HTTP::LogHandler
...
-      @log.info { "#{req.remote_address || "-"} - #{req.method} #{req.resource} #{req.version} - #{res.status_code} (#{elapsed_text})" }
+
+      addr =
+        case remote_address = req.remote_address
+        when nil
+          "-"
+        when Socket::IPAddress
+          remote_address.address
+        else
+          remote_address.to_s
```
Comment (straight-shoota) [inline src/http/server/handlers/log_handler.cr:35]: That to_s is unncessary. The value is stringified in the interpolation.
```suggestion
          remote_address
```

**Comment by waj** [comment]
I just disabled some specs in win32, mostly for HTTP. I'm confused about the `require "http"` within `src/mime/media_type.cr` but I removed it and then I cannot run some individual specs. So probably a task for a separate PR.

**Comment by straight-shoota** [comment]
Disabling that many specs which worked before this PR is not a good idea. We can just make `remote_address = nil` on win32 for now and it should be fine.

**Comment by waj** [comment]
Ok, I’ll try that

**Comment by waj** [comment]
@straight-shoota the specs in win32 are back. I left `String?` as the type for `remote_address` in Windows for now.

**Comment by straight-shoota** [comment]
Why `String`? Can't it just be nil on win32?

Public APIs should not have different types depending on the platform.

**Comment by waj** [comment]
I put `String` to avoid disabling more specs. Nobody is going to use `HTTP` on Windows yet, because sockets are not working yet, AFAIK.

**Comment by RX14** [comment]
Types cannot change based on platform.

**Pull request merged**